File: python/course-repo/solution/06-same-code-runs-anywhere/src/webhooks/workflow.py
from datetime import timedelta

from temporalio import workflow

# This Workflow uses the SAME deliver_webhook function from activities.py
# that the standalone caller uses. No copies, no rewrites - write the
# Activity once, compose it into a multi-step Workflow when the job grows.
# Both imports go through imports_passed_through so the Python SDK
# sandbox doesn't reload activities.py / shared.py on every workflow task.
with workflow.unsafe.imports_passed_through():
    from .activities import deliver_webhook
    from .shared import WebhookDelivery


@workflow.defn
class WebhookWorkflow:
    """Thin Workflow that calls deliver_webhook as a step.

    Demonstrates that the same Activity code runs as a top-level job
    (client.execute_activity) or as a Workflow step (workflow.execute_activity).
    """

    @workflow.run
    async def run(self, req: WebhookDelivery) -> int:
        return await workflow.execute_activity(
            deliver_webhook,
            req,
            start_to_close_timeout=timedelta(seconds=10),
        )

File: typescript/course-repo/exercise/06-same-code-runs-anywhere/src/workflow.ts
import { proxyActivities } from '@temporalio/workflow';
import type { WebhookDelivery } from './shared';
import type * as activities from './activities';

// This Workflow uses the SAME deliverWebhook function from activities.ts
// that the standalone caller uses. No copies, no rewrites.
const { deliverWebhook } = proxyActivities<typeof activities>({
  startToCloseTimeout: '10 seconds',
});

export async function webhookWorkflow(req: WebhookDelivery): Promise<number> {
  return await deliverWebhook(req);
}
